File: R/read_feh.R

library(tidyverse)


#' Returns a vector of field types
#'
#' @param lines Character vector of codebook lines
#' @return A vector of field types
get_record_types <- function(lines)
{
    scalar_line_numbers <- grep("[[:space:]]*V-[[:space:]]*[[:alnum:]]+[[:space:]]+", lines)
    array_line_numbers <- grep("[[:space:]]*V-[[:space:]]*[[:alnum:]]+\\(", lines)
    rec_type <- character(length(lines))
    rec_type[scalar_line_numbers] <- "scalar"
    rec_type[array_line_numbers] <- "array"
    return(rec_type)
}

#' Parse codebook lines into a dataframe of names, types and array ranges
#'
#' @param lines Character vector of codebook lines
#' @return A dataframe with variables: name, type, start, stop
parse_codebook_lines <- function(lines)
{
    per_names <- sub("[[:space:]]*V-[[:space:]]*([[:alnum:]]+)[[:space:](].*", "\\1", lines)
    rec_type <- get_record_types(lines)
    array_year_range <- sub("[[:space:]]*V-[[:space:]]*[[:alnum:]]+[[:space:]]+.*","", lines)
    array_year_range <- sub("[[:space:]]*V-[[:space:]]*[[:alnum:]]+\\(([[:digit:]]+\\-[[:digit:]]+)\\).*","\\1", array_year_range)
    array_year_start <- as.integer(sub("([[:digit:]]+)\\-.*","\\1", array_year_range))
    array_year_end   <- as.integer(sub("[[:digit:]]+\\-([[:digit:]]+).*","\\1", array_year_range))
    return(data.frame(name=per_names, type=rec_type, start=array_year_start, end=array_year_end))
}

#fam_df <- parse_codebook_lines(fam_rec_lines)

#' Parse person record codebook
#'
#' @param fname The codebook file name
#' @return A dataframe with variables: name, type, start, stop
parse_person_codebook <- function(fname)
{
    codebook_text <- readLines(fname)

    fam_rec_start <- grep("[[:space:]]*RECORD- FAMILY", codebook_text)
    per_rec_start <- grep("[[:space:]]*RECORD- PERSON", codebook_text)
    endvars <- grep("[[:space:]]*ENDVARS", codebook_text)

    if(!(endvars[1] > fam_rec_start & endvars[1] < per_rec_start & endvars[2] > per_rec_start)) {
        stop("ERROR: Codebook invalid")
    }

    per_rec_lines <- codebook_text[(per_rec_start+1):(endvars[2]-1)]
    per_rec_lines <- grep("[[:space:]]*V-", per_rec_lines, value = T)

    return(parse_codebook_lines(per_rec_lines))
}

#' Parse family record codebook
#'
#' @param fname The codebook file name
#' @return A dataframe with variables: name, type, start, stop
parse_family_codebook <- function(fname)
{
    codebook_text <- readLines(fname)

    fam_rec_start <- grep("[[:space:]]*RECORD- FAMILY", codebook_text)
    per_rec_start <- grep("[[:space:]]*RECORD- PERSON", codebook_text)
    endvars <- grep("[[:space:]]*ENDVARS", codebook_text)

    if(!(endvars[1] > fam_rec_start & endvars[1] < per_rec_start & endvars[2] > per_rec_start)) {
        stop("ERROR: Codebook invalid")
    }

    fam_rec_lines <- codebook_text[(fam_rec_start+1):(endvars[1]-1)]
    fam_rec_lines <- grep("[[:space:]]*V-", fam_rec_lines, value = T)

    return(parse_codebook_lines(fam_rec_lines))
}

df_to_col_names <- function(df)
{
    rec_struct <- as.character(df$name)

    arrays_df <- df %>%
        filter(type=="array") %>%
        select(name, start, end)

    rec_struct <- c(rec_struct,
                    unlist(apply(arrays_df, 1, function(x) paste0(x[1], x[2]:x[3]))))
}

get_col_names = function(codebook)
{
    per_df <- parse_person_codebook(codebook)
    col_names <- df_to_col_names(per_df)
    return(col_names)
}


#' Reads DYNASIM codebook and FEH persons file
#'
#' @param codebookfile the path to an FEH codebook
#' @param fehfile the path to an FEH persons file
#' @param columns(optional) character vector of column names to keep; all if NULL (default)
#' @param obs_count the number of observations to read; all if NA (default)
read_feh_cbk = function(codebookfile, fehfile, columns=NULL, obs_count=NA)
{

    # Read the codebook
    rec_struct <- get_col_names(codebookfile)
    rec_len <- length(rec_struct)

    # FEH file size
    feh_file_size <- file.size(fehfile)

    # FEH number of records
    rec_num <- feh_file_size/4/rec_len

    # Number of observations to read
    if(is.na(obs_count)) {
        obs_count <- rec_num
    }
    else if(obs_count<=0 || obs_count>rec_num) {
        stop("Invalid number of observations obs_count=", obs_count)
    }

    chunk_size = 50000

    fehcon = file(fehfile, "rb")
    obs_read=0
    while(obs_read<obs_count) {

        read_count = pmin(obs_count-obs_read, chunk_size)
        read_len = read_count * rec_len

        # Read raw FEH data
        cat("Reading next ", format(read_count, width=6), " observations...")
        rawdata <- readBin(fehcon, "integer", read_len, 4)

        # Convert raw data into a matrix
        feh_mat <- matrix(rawdata, nrow=read_count, ncol=rec_len, byrow = T)

        colnames(feh_mat) <- rec_struct

        # Convert to data frame
        df <- as.data.frame(feh_mat)

        if(!is.null(columns)) {
            df = select(df, all_of(columns))
        }

        if(obs_read == 0) {
            feh_df = df
        }
        else {
            feh_df = bind_rows(feh_df, df)
        }
        obs_read = obs_read + read_count
        cat("total =", format(obs_read, width=8, scientific=FALSE, big.mark=','), "\n")
    }
    close(fehcon)
    return(feh_df)
}

#' Converts DYNASIM FEH output data from wide to long format.
#' It also changes the column names into lower case.
#'
#' @param data DYNASIM FEH dataset as returned by read_feh
#'
feh_wide_to_long = function(data)
{
    # Time-series names
    tsnames = tolower(unique(gsub("\\d{4}", "", grep("[A-Z]+\\d{4}", keep_cols, value = TRUE))))
    match_pat = paste(paste0(tsnames, ".*"), collapse="|")
    names_pat = paste0("(", paste(tsnames, collapse = "|"), ")(\\d+)")

    df = data |>
        rename_with(tolower) |>
        pivot_longer(
            matches(match_pat),
            names_to=c('.value', 'year'),
            names_pattern=(names_pat)
        ) |>
        mutate(year=as.integer(year))

    return(df)
}

#' Reads from an open FEH header file and returns a record structure
#' 
#' @param file a connection to a header file pointing to a beginning of a header
#' 
make_record_struct <- function(file) {
  
  # Read binary data from file
  bin_data <- readBin(file, what = "raw", n = 44)
  
  # Unpack binary data
  nsat <- as.integer(rawToChar(bin_data[3:12]))
  nmat <- as.integer(rawToChar(bin_data[13:22]))
  rlen <- as.integer(rawToChar(bin_data[23:32]))
  nmts <- as.integer(rawToChar(bin_data[33:42]))
  
  # Create result dictionary
  res <- list(
    nsat = nsat,
    nmat = nmat,
    rlen = rlen,
    nmts = nmts
  )
  
  # Read variable names
  res$names <- vector(mode = "character", length = res$nsat)
  for (i in seq_len(res$nsat)) {
    res$names[i] <- readChar(file, 8, useBytes = TRUE)
  }
  
  # Read MTS variables if nmts > 0
  if (res$nmts > 0) {
    readBin(file, "integer", size=2, n = 1) # Skip 2 bytes
    res$mtsnm <- numeric(res$nmts)
    res$mtsly <- numeric(res$nmts)
    res$mtshy <- numeric(res$nmts)
    res$mtsp <- numeric(res$nmts)
    for (i in seq_len(res$nmts)) {
      res$mtsnm[i] <- as.integer(readBin(file, "integer", size=4, n = 1))
    }
    for (i in seq_len(res$nmts)) {
      res$mtsly[i] <- as.integer(readBin(file, "integer", size=4, n = 1))
    }
    for (i in seq_len(res$nmts)) {
      res$mtshy[i] <- as.integer(readBin(file, "integer", size=4, n = 1))
    }
    for (i in seq_len(res$nmts)) {
      res$mtsp[i] <- as.integer(readBin(file, "integer", size=4, n = 1))
    }
  }
  
  return(res)
}

#' Creates a vector of column names from a header record
#' 
#' @param rec record structure returned by make_record_struct
#' 
make_col_names <- function(rec) {
  # Start with all variable names--these are scalar variables
  names <- trimws(rec$names)
  
  # Now add MTS variables
  for (i in seq_len(rec$nmts)) {
    # MTS variable name
    mtsname <- names[rec$mtsnm[i]]
    
    # Combine MTS name with all the years from its range
    mtsnames <- paste0(mtsname, rec$mtsly[i]:rec$mtshy[i])
    
    names <- c(names, mtsnames)
  }
  
  # Return the dtype list
  return(names)
}

#' Reads an FEH header file and returns its record structures
#' 
#' @param hdrfile the path to a header file
#' 
read_header = function(hdrfile)
{
  
  file <- file(hdrfile, "rb")
  year = readBin(file, what="raw", n=10)
  year = as.integer(rawToChar(year))
  
  famrec <- make_record_struct(file)
  perrec <- make_record_struct(file)
  close(file)
  res = list(
    famrec=famrec,
    perrec=perrec
  )
  return(res)
}



#' Reads FEH header and data files
#' 
#' File names must follow the following convention:
#'  header: dynasipp_header_[even|odd].dat
#'  data:   dynasipp_[person|family]_[even|odd].dat
#'
#' @param fehdir the folder containing the header and data files
#' @param even (optional) read even-year files; otherwise, read odd-year files
#' @param person (optional) 
#' @param columns(optional) character vector of column names to keep; all if NULL (default)
#' @param obs_count the number of observations to read; all if NA (default)
read_feh = function(fehdir, even=TRUE, person=TRUE, columns=NULL, obs_count=NA)
{
  year_sfx = ifelse(even,  'even',  'odd')
  type_sfx = ifelse(person,'person','family')
  
  hdrfile = paste0(fehdir, '/dynasipp_header_', year_sfx, '.dat')
  datfile = paste0(fehdir, '/dynasipp_', type_sfx, '_', year_sfx, '.dat')
  
  stopifnot(file.exists(hdrfile))
  stopifnot(file.exists(hdrfile))

  # Read the header
  hdr = read_header(hdrfile)  
  if( person ) {
    rec=make_col_names(hdr$perrec)
  } 
  else {
    rec=make_col_names(hdr$famrec)
  }
  rec_len <- length(rec)
  
  # FEH file size
  feh_file_size <- file.size(datfile)
  
  # FEH number of records
  rec_num <- feh_file_size/4/rec_len
  
  # Number of observations to read
  if(is.na(obs_count)) {
    obs_count <- rec_num
  }
  else if(obs_count<=0 || obs_count>rec_num) {
    stop("Invalid number of observations obs_count=", obs_count)
  }
  
  chunk_size = 50000
  
  fehcon = file(datfile, "rb")
  obs_read=0
  while(obs_read<obs_count) {
    
    read_count = pmin(obs_count-obs_read, chunk_size)
    read_len = read_count * rec_len
    
    # Read raw FEH data
    cat("Reading next ", format(read_count, width=6), " observations...")
    rawdata <- readBin(fehcon, "integer", read_len, 4)
    
    # Convert raw data into a matrix
    feh_mat <- matrix(rawdata, nrow=read_count, ncol=rec_len, byrow = T)
    
    colnames(feh_mat) <- rec
    
    # Convert to data frame
    df <- as.data.frame(feh_mat)
    
    if(!is.null(columns)) {
      df = select(df, all_of(columns))
    }
    
    if(obs_read == 0) {
      feh_df = df
    }
    else {
      feh_df = bind_rows(feh_df, df)
    }
    obs_read = obs_read + read_count
    cat("total =", format(obs_read, width=8, scientific=FALSE, big.mark=','), "\n")
  }
  close(fehcon)
  return(feh_df)
}


example=FALSE
test=FALSE
if(example) {
    fehdir = "S:/damir/run_1004/"
    fehfile <- paste0(fehdir, "dynasipp_person_even.dat")
    codebookfile = paste0(fehdir, "codebook_2100ds.sipp2006")
    keep_cols = c('PERNUM', 'DOBY', 'SEX', 'RACE', 'ETHNCTY', 'YEARDIED', paste0('GRADECAT', 2006:2100))
    df = read_feh(codebookfile, fehfile, columns=keep_cols)
}
if(test) {
  
  # Compare data read by read_feh_cbk and read_feh  
  
  df1 = read_feh(
    "X:/FEHOutput/run_01006/base_v5",
    obs_count=1e4
  )
  
  df2 = read_feh_cbk(
    "C:/Users/dcosic/Dynasim/Dynasim-babybonds/Dynasim-core/run/codebook_2100ds.sipp2006",
    "X:/FEHOutput/run_01006/base_v5/dynasipp_person_even.dat",
    obs_count=1e4
  )
  
  ncols = dim(df1)[2]
  
  testpass=TRUE
  for(i in seq_len(ncols)) {
    if( ! all(df1[i]==df2[i], na.rm=TRUE) ) {
      testpass = FALSE
      print(paste0("Column ", i, " differs"))
    }
  }
  if(testpass) {
    print("Test passed")
  }
}  

File: Python/read_feh.py
"""
Functions for reading DYNASIM-FEH output files.

DYNASIM FEH produces three files:

    - header file,
    - family file, and
    - person file.


The header file contains specifications of both family and person
records. The function that reads DYNASIM data is read_feh_data_file. 
It requires paths to a header file and to one of the two data files.
"""

import struct
import numpy as np
import io

def make_record_dict(file:io.BufferedReader, sample:str):
    """Reads a section of a header file and creates a record dictionary

    Args:
        file (io.BufferedReader): a file object pointing to the 
        beginning of a header section.

    Returns:
        dict: A dictionary with the following elements:
            nsat: total number of variables
            nmat: NA
            rlen: NA
            nmts: number of micro-time-series (MTS) variables
            names: list of variable names
            mtsnm: list of MTS variable names as indices to names list
            mtsly: list of first years in MTS variables
            mtshy: list of last years in MTS variables
            mtsp:  list of offsets to MTS variables
    """

    res = {}

    if sample == "output":
        nsat, nmat, rlen, nmts = struct.unpack('2x10s10s10s10s2x', file.read(44))
        #print(int.from_bytes(nsat, byteorder="little", signed=True))
        res['nsat'] = int(nsat)
        res['nmat'] = int(nmat)
        res['rlen'] = int(rlen)
        res['nmts'] = int(nmts)

    elif sample == "input":
        unused = np.fromfile(file, dtype=np.uint8, count=2)
        frame = np.fromfile(file, dtype=np.int32, count=4)
        unused = np.fromfile(file, dtype=np.uint8, count=2)
        res = {
        "nsat": frame[0],
        "nmat": frame[1],
        "rlen": frame[2],
        "nmts": frame[3]
        }

    # Create names list
    res['names'] = [''] * res['nsat']
    for i in range(res['nsat']):
        res['names'][i] = str(file.read(8), 'utf-8')

    # If there is at least 1 variable, initialize 0 vectors to be filled later
    if res['nmts'] > 0:
        file.read(2)
        res['mtsnm'] = [0]*res['nmts']
        res['mtsly'] = [0]*res['nmts']
        res['mtshy'] = [0]*res['nmts']
        res['mtsp']  = [0]*res['nmts']

        # Populate mtsp, mtsly, mtshy, mtsp 0 vectors, read bytes as integers
        for i in range(res['nmts']):
            res['mtsnm'][i] = struct.unpack('i', file.read(4))[0]-1

        for i in range(res['nmts']):
            res['mtsly'][i] = struct.unpack('i', file.read(4))[0]

        for i in range(res['nmts']):
            res['mtshy'][i] = struct.unpack('i', file.read(4))[0]

        for i in range(res['nmts']):
            res['mtsp'][i] = struct.unpack('i', file.read(4))[0]

    return res

def make_rec_dtype(rec:dict):
    """Creates a dtype object based on a record dictionary

    Args:
        rec (dict): a dictionary that describes a record

    Returns:
        dtype: A dtype object for a numpy structured array
    """
    
    # Start with all variable names--these are scalar variables
    names = [n.strip() for n in rec['names']]

    # Now add MTS variables
    for i in range(rec['nmts']):

        # MTS variable name
        mtsname = names[rec['mtsnm'][i]]

        # Combine MTS name with all the years from its range
        mtsnames = [
            mtsname + str(y) for y in 
            range(rec['mtsly'][i], rec['mtshy'][i]+1)
            ]

        names.extend(mtsnames)

    # Create a dtype object for numpy structured arrays
    rectype = np.dtype({'names':names, 'formats': ['i4']*len(names)})

    return rectype

def read_header_file(filename:str):
    """Reads a DYNASIM header file and returns record dictionaries.
       Presumes the first 4 bytes of the starting sample file are '2006' and first 10 bytes of 
       output sample are '2060'. If neither condition satisfied, error is raised.

    Args:
        filename (str): the path to a DYNASIM header file

    Returns:
        tuple: (year, family-record dict, person-record dict)
    """

    # Open header file
    with open(filename, 'rb') as file:
        # Get the first 10 bytes
        first_10_bytes = file.read(10)

        # Check if the first 4 bytes of the first 10 represent the year 2006
        if int.from_bytes(first_10_bytes[:4], byteorder="little", signed=True) == 2006:
            sample = 'input'
            year = int.from_bytes(first_10_bytes[:4], byteorder="little", signed=True)
            # Go back to the 4th byte to ensure the file is in the correct place
            file.seek(4)
        
        # Check if the first 10 bytes represent the year 2060
        elif int(first_10_bytes) == 2060:
            # If output sample, need to move 6 more bytes
            sample = 'output'
            year = int(first_10_bytes)
        
        # If neither condition is met, raise an error
        else:
            raise ValueError( 'Invalid year in file header:\n'
                              'The first 4 bytes should contain 2006 or 2060 to identify data.\n'
                              f'First ten bytes, raw: {first_10_bytes}' )

        # Make family-record dictionary
        famrec = make_record_dict(file, sample = sample)

        # Make person-record dictionary
        perrec = make_record_dict(file, sample = sample)

    return year, famrec, perrec

def read_feh_data_file(
        header_file:str, 
        data_file:str,
        file_type:str='person', 
        count:int=-1):
    """Reads a DYNASIM data file

    Args:
        header_file (str): the path to a DYNASIM header file
        data_file (str): the path to a DYNASIM data file
        file_type (str): 'person' or 'family' file. Defaults to 
            'person'.
        count (int, optional): number of records to read, all if -1. 
            Defaults to -1.

    Returns:
        numpy structured array: data
    """

    year, famrec, perrec = read_header_file(header_file)
    
    if file_type == 'person':
        rectype = make_rec_dtype(perrec)

    elif file_type == 'family':
        rectype = make_rec_dtype(famrec)

    else:
        raise ValueError(f"file_type can be 'person' or 'family' but not {file_type}")
    
    return np.fromfile(data_file, dtype=rectype, count=count)


def feh_wide_to_long(widearr):
    """Creates a long-format array for longitudinal variables in widearr

    Args:
        widearr (np.array): structured numpy array with longitudinal variables and "PERNUM"

    Returns:
        structured numpy array: long-format array with longitudinal variables from widearr
    """

    # List of years in the longitudinal variables
    years = np.unique([int(x[-4:]) for x in widearr.dtype.names if x[-4:].isdigit()])

    # The list of longitudinal variables in the wide format
    mtswide = [x for x in widearr.dtype.names if x[-4:].isdigit()]

    # The list of longitudinal variable names
    mtslong = set([x[:-4] for x in mtswide])
    mtslong = [x.lower() for x in mtslong]

    # The list of variable names for the long array
    longnames = ['year', 'pernum']
    longnames.extend(mtslong)

    # Dimensions
    nper = widearr.shape[0]
    nmtsw = len(mtswide)
    nyears = len(years)
    nrow = nper * nyears

    # Create an empty array
    rectype = np.dtype({'names':longnames, 'formats': ['i4']*len(longnames)})
    longarr = np.zeros(nrow, rectype)

    # Fill in pernum and year
    longarr['pernum'] = np.tile(widearr['PERNUM'], nyears)
    longarr['year']   = np.repeat(np.array(years), nper)

    # For each year, fill in the rest of the variables
    for y, year in enumerate(years):
        yl = y*nper
        yh = (y+1)*nper
        yvarsw = [var for var in mtswide if str(year) in var]
        yvarsl = [x[:-4].lower() for x in yvarsw]
        longarr[yvarsl][yl:yh] = widearr[yvarsw]

    return longarr

def print_offsets(d):

    print("offsets:", [d.fields[name][1] for name in d.names])

    print("itemsize:", d.itemsize)

def structured_shape(x):
    if len(x.dtype) > 0:
        return list(x.shape) + [len(x.dtype)]
    else:
        return x.shape

if __name__ == '__main__':

    sample_dic = {
        "input": {
            "header_file": '../data/starting-sample/v2/dynasipp_HEADER.dat',
            "person_file": '../data/starting-sample/v2/dynasipp_PERSON.dat',
            "family_file": '../data/starting-sample/v2/dynasipp_FAMILY.dat'
        },
        "output": {
            "header_file": '../data/output/run-1006-baseline/base-v8/dynasipp_header_even.dat',
            "person_file": '../data/output/run-1006-baseline/base-v8/dynasipp_person_even.dat',
            "family_file": '../data/output/run-1006-baseline/base-v8/dynasipp_family_even.dat'           
        }
    }

    sample_type = "input"

    header_file = sample_dic[sample_type]["header_file"]
    person_file = sample_dic[sample_type]["person_file"]
    family_file = sample_dic[sample_type]["family_file"]

    perdata = read_feh_data_file(header_file, person_file, file_type='person', count=10)
    famdata = read_feh_data_file(header_file, family_file, file_type='family', count=10)

    longarr = feh_wide_to_long(perdata)
    print(perdata.dtype)
    #print(perdata)
    #print(longarr)
    # pd = perdata.dtype
    # # Flags
    # print(perdata.flags)
    # # Convert to record array
    # for rec in recordarr:
    #     print(rec.AGE)
    # print(perdata.shape)
    # print(structured_shape(perdata))
    # Convert to record array
    # recordarr = np.rec.array(perdata)
    # print(recordarr[:].AGE)
    # print(pd.names[0:10])
    # for x in pd.names:
    #     print(x[-4:])